src/engine/runtime.js

```diff
@@ -802,13 +805,14 @@ class Runtime extends EventEmitter {
 
             // If no fields are present, check inputs (horizontal blocks)
             if (Object.keys(hatFields).length === 0) {
+                hatFields = {}; // don't overwrite the block's actual fields list
                 const hatInputs = blocks.getInputs(block);
                 for (const input in hatInputs) {
                     if (!hatInputs.hasOwnProperty(input)) continue;
                     const id = hatInputs[input].block;
                     const inpBlock = blocks.getBlock(id);
                     const fields = blocks.getFields(inpBlock);
-                    hatFields = Object.assign(fields, hatFields);
+                    Object.assign(hatFields, fields);
```

Comment:
My guess is the original line here was written thinking that it would act like `hatFields = Object.assign({}, fields, hatFields)`. This new code should be equivalent except that it throws away fewer intermediate objects in cases where there are multiple inputs. In my opinion it's also easier to understand: in particular, it's now obvious that `Object.assign` is expected to modify one of the objects.